crates/core/src/routing_info/mod.rs

```diff
@@ -0,0 +1,219 @@
+    partition_to_node_mappings: Arc<ArcSwap<PartitionToNodesRoutingTable>>,
+}
+
+impl PartitionRouting {
+    /// Look up a suitable node to answer requests for the given partition.
+    pub fn get_node_by_partition(&self, partition_id: PartitionId) -> Option<NodeId> {
+        self.partition_to_node_mappings
+            .load()
+            .inner
+            .get(&partition_id)
+            .copied()
+    }
+
+    /// Request a refresh without waiting for it to complete. This is useful when the caller
+    /// discovers via some other mechanism that the local view might be outdated - for example, when
+    /// a request to a node previously returned by `get_node_by_partition` fails with a response
+    /// that indicates that this routing information is no longer valid.
+    pub async fn request_refresh(&self) {
+        self.sender
+            .send(Command::SyncRoutingInformation(TargetVersion::Latest))
+            .await
+            .expect("Failed to send refresh request");
+    }
+}
+
+struct PartitionToNodesRoutingTable {
+    /// A mapping of partition IDs to node IDs that are believed to be authoritative for that
+    /// serving requests for that partition.
+    inner: HashMap<PartitionId, NodeId, Xxh3Builder>,
+}
+
+/// Task to refresh the routing information, periodically or on-demand. A single
+/// instance of this component per node ensures that we don't get a stampeding
+/// herd of queries to the source of truth.
+//
+// Implementation note: the sole authority for node-level routing is currently the metadata store,
+// and in particular, the cluster scheduling plan. This will change in the future so avoid leaking
+// implementation details or assumptions about the source of truth.
+pub struct PartitionRoutingRefresher {
+    sender: CommandSender,
+    receiver: CommandReceiver,
+    metadata_store_client: MetadataStoreClient,
+    inflight_refresh_task: Option<TaskHandle<anyhow::Result<()>>>,
+    inner: Arc<ArcSwap<PartitionToNodesRoutingTable>>,
+}
+
+impl PartitionRoutingRefresher {
+    pub fn new(metadata_store_client: MetadataStoreClient) -> Self {
+        let (sender, receiver) = mpsc::channel(1);
+        Self {
+            metadata_store_client,
+            receiver,
+            sender,
+            inflight_refresh_task: None,
+            inner: Arc::new(ArcSwap::new(Arc::new(PartitionToNodesRoutingTable {
+                inner: HashMap::default(),
+            }))),
+        }
+    }
+
+    /// Get a handle to the partition-to-node routing table.
+    pub fn partition_routing(&self) -> PartitionRouting {
+        PartitionRouting {
+            sender: self.sender.clone(),
+            partition_to_node_mappings: self.inner.clone(),
+        }
+    }
+
+    async fn run(mut self) -> anyhow::Result<()> {
+        debug!("Metadata manager started");
+
+        let update_interval = Configuration::pinned()
+            .common
+            .metadata_update_interval
+            .into();
```

Comment:
Wasn't sure what to pick for this. It seems reasonable to start with; it might be a nice optimisation to add a `get_if_newer_version` to the metadata store to make it cheaper to poll. Eventually we should gossip/push this info to nodes without needing to poll all the time.

Reply:
Wondering whether the periodic update is really required if users of this struct refresh explicitly on outdated leader information.

Reply:
I don't believe we have any explicit mechanism today to provide outdated feedback; I exposed the refresh method just in case, but let's keep this for now please. Happy to reduce the refresh interval to something like 3-5x the heartbeat interval initially.

Reply:
Let's see whether the load on the metadata store will be too high. Something to keep an eye out for.

Reply:
Yeah! Not entirely certain what signals to monitor that will tell us that this is a problem, before it becomes an actual problem, if that makes sense. Probably some metrics around metadata store service?

I'm pretty keen on the `get_if_version_newer` optimisation - I think basically any backing store we would choose for metadata should easily support this feature, and worst case, just fall back on always doing a full read if it can't. At least saves some bytes over the wire.